packages/ember-data/lib/system/model/internal-model.js

```diff
@@ -0,0 +1,696 @@
+  this.dataHasInitialized = false;
+  this._deferredTriggers = [];
+  this._data = {};
+  this._attributes = Ember.create(null);
+  this._inFlightAttributes = Ember.create(null);
+  this._relationships = {};
+  this.currentState = RootState.empty;
+  /*
+    implicit relationships are relationship which have not been declared but the inverse side exists on
+    another record somewhere
+    For example if there was
+    ```
+      App.Comment = DS.Model.extend({
+        name: DS.attr()
+      })
+    ```
+    but there is also
+    ```
+      App.Post = DS.Model.extend({
+        name: DS.attr(),
+        comments: DS.hasMany('comment')
+      })
+    ```
+
+    would have a implicit post relationship in order to be do things like remove ourselves from the post
+    when we are deleted
+  */
+  this._implicitRelationships = Ember.create(null);
+  var model = this;
+  //TODO Move into a getter for better perf
+  this.eachRelationship(function(key, descriptor) {
+    model._relationships[key] = createRelationshipFor(model, descriptor, model.store);
+  });
+};
+
+InternalModel.prototype = {
+  isEmpty: retrieveFromCurrentState('isEmpty'),
+  isLoading: retrieveFromCurrentState('isLoading'),
+  isLoaded: retrieveFromCurrentState('isLoaded'),
+  isDirty: retrieveFromCurrentState('isDirty'),
+  isSaving: retrieveFromCurrentState('isSaving'),
+  isDeleted: retrieveFromCurrentState('isDeleted'),
+  isNew: retrieveFromCurrentState('isNew'),
+  isValid: retrieveFromCurrentState('isValid'),
+  dirtyType: retrieveFromCurrentState('dirtyType'),
+
+  constructor: InternalModel,
+  materializeRecord: function() {
+    // lookupFactory should really return an object that creates
+    // instances with the injections applied
+    this.record = this.type._create({
+      id: this.id,
+      store: this.store,
+      container: this.container
+    });
+    this.record._internalModel = this;
+    this._triggerDeferredTriggers();
+  },
+
+  recordObjectWillDestroy: function() {
+    this.record = null;
+  },
+
+
+  deleteRecord: function() {
+    this.send('deleteRecord');
+  },
+
+  save: function() {
+    var promiseLabel = "DS: Model#save " + this;
+    var resolver = Ember.RSVP.defer(promiseLabel);
+
+    this.store.scheduleSave(this, resolver);
+    return resolver.promise;
+  },
+
+  startedReloading: function() {
+    this.isReloading = true;
+    if (this.record) {
+      set(this.record, 'isReloading', true);
+    }
+  },
+
+  finishedReloading: function() {
+    this.isReloading = false;
+    if (this.record) {
+      set(this.record, 'isReloading', false);
+    }
+  },
+
+  reload: function() {
+    this.startedReloading();
+    var record = this;
+    var promiseLabel = "DS: Model#reload of " + this;
+    return new Promise(function(resolve) {
+      record.send('reloadRecord', resolve);
+    }, promiseLabel).then(function() {
+      record.didCleanError();
+      return record;
+    }, function(reason) {
+      record.didError();
+      throw reason;
```

Comment:
good point, can you open a separate issue for that, i just copied the method around